src/fragments/Chat.tsx

```diff
@@ -13,25 +14,52 @@ import ChatMessageInput from './ChatMessageInput';
 import ChatMessages from './ChatMessages';
 
 const Chat: React.FC = () => {
-  const { channel, startChatSession, prependToMessages } =
+  const { channel, startChatSession, prependToMessages, currentUser } =
     useContext(ChatAppContext);
-
+  const [typingUsers, setTypingUsers] = useState<User[]>([]);
+  const [numUserTyping, setNumTyping] = useState(0);
+  const [typingUser, setTypingUser] = useState<User | null>(null);
   useEffect(() => {
     if (!channel) {
       return;
     }
-
-    const session = startChatSession(channel, (message) => {
-      prependToMessages([message]);
-    });
+    const session = startChatSession(
+      channel,
+      (message: Message) => {
+        prependToMessages([message]);
+      },
+      (user: User) => {
+        if (currentUser?.id !== user.id) {
+          setTypingUser(user);
+          setTypingUsers((typingUsers) => [...typingUsers, user]);
+          if (numUserTyping < 5) {
+            setNumTyping(numUserTyping + 1);
```

Comment:
Why not use `typingUsers.length` to determine the number of typing users? Removes the need for extra state management.